### ISSUE TITLE
User properties doesn't clear anymore

### ISSUE THREAD

**Issue by mquigley-ck** [amplitude/Amplitude-iOS#63]
A recent update broke the `setUserProperties replace` method, where the `replace` parameter is ignored.
https://github.com/amplitude/Amplitude-iOS/commit/540a8e2c603fc7532a1ed421fda64cfb313126df#diff-2777663457a583bbf669698eaadec79bR1016

 On top of that, the README.md still describes using `replace`:

> To replace any existing user properties with a new set:
> NSMutableDictionary *userProperties = [NSMutableDictionary dictionary];
> [userProperties setValue:@"VALUE_GOES_HERE" forKey:@"KEY_GOES_HERE"];
> [[Amplitude instance] setUserProperties:userProperties replace:YES];

Previously, the following would clear user properties:

    [[Amplitude instance] setUserProperties:@{} replace:YES];

The most onerous problem that happened is that the contract of a public SDK changed without warning or notice. It break contracts of public SDKs. The function was left in a public SDK so that it compiles, but quietly ignores the parameter with no warning whatsoever.

Regardless of the broken contract, AFAICT the new API doesn't have the ability to to "clear all" properties. It appears you have to know which properties existed before, which may not be possible.

**Comment by djih** [comment]
Hi @mquigley-ck, it's a little complicated but hopefully I can clarify for you.

You might be misunderstanding what the replace was doing in the first place, which is understandable because the naming is a little confusing. Before [v3.2.0](https://github.com/amplitude/Amplitude-iOS/releases/tag/v3.2.0), userProperties dict was stored in the SDKs memory: https://github.com/amplitude/Amplitude-iOS/blob/v3.1.1/Amplitude/Amplitude.m#L69

This dict would be sent with every event sent via logEvent https://github.com/amplitude/Amplitude-iOS/blob/v3.1.1/Amplitude/Amplitude.m#L473. So what setUserProperties with replace:YES would do is replace the NSDictionary in the SDK's memory with a new one, and send that new one on every logEvent.

What you probably didn't know is that Amplitude's event servers propagate user properties. For example if a user logs event 1 with user properties {'a': 1, 'b':2}, and then logs events 2, 3, 4 without any user properties, then the server automatically adds properties {'a': 1, 'b':2} to events 2, 3, 4. Even if you did setUserProperties with an empty NSDictionary, that would not clear those properties from the user. In addition if you then tried doing setUserProperties replace:Yes with properties {'c': 3, 'd': 4}, then this would replace the NSDictionary on the SDK (and set it to {'c': 3, 'd': 4}), but the server would actually combine all of the previous properties, and you would see {'a'...'b'...'c'...'d'...}.

The important thing to note is that because the server already propagates the user properties, there is no reason to send them with every event. You end up transmitting unnecessary duplicate data. You only need to communicate to the server when the properties change. For that reason in v3.2.0 we no longer store userproperties on the SDK. When you do a setUserProperty, the updates are immediately communicated to the server. That means that replace does nothing as there is nothing to replace. We decided to keep the method signature so legacy apps don't break.

This also means that before v3.1.0, there was no way at all to clear user properties. Setting userProperties to an empty dictionary with replace:Yes did not do that. No contract was broken because the ability to clear all user properties was never in the contract to begin with.

In v3.2.0 we added the unset operator allowing you to unset specific properties. https://github.com/amplitude/Amplitude-iOS#user-property-operations. That does, however, require you to know that property you want to be clearing. 

That being said, it wouldn't be too much work for us to implement a user property operation `clear` or something, if that's what CreditKarma needs. :)
